### PR TITLE
[DDW-410] Cancel running fee estimation api requests once new request is made

### PR DESCRIPTION
**Pull request by yakovkaravelov**
This PR canceled running fee estimation api requests once new request is made
[Jira Ticket](https://jira.iohk.io/browse/DDW-410)

## Todos
- [x] Cancel running fee estimation api requests once new request is made

## Testing Checklist
- [Slack QA thread]()

### App
- [x] Type numbers in ADA field on the send screen and make sure that fee estimation api requests are canceled whenever typed numbers are changed so that fee calculation gets more faster and up-to-date with latest input
- [x] Make sure app is working properly

## Testing Summary
- Please refer the snapshot and try testing the scenario described
image

For build 15645:
**Pending**
1 - When the value changes, the "next" button is still enabled for a fraction of a second. If you quickly press enter after changing the value, you'll get to the confirmation dialog with no fees. Check[ this video.](https://drive.google.com/file/d/1nCl8MfzIgsnidmLab_gXS7m5Y1uJIWAo/view?usp=sharing)
Eventually, the correct fees are displayed in the dialog and the user can submit the transaction, but I'm wondering if this can be improved to avoid this awkward scenario.

**Fixed**
1 - When switching the value, sometimes the "calculating fees" message disappears (for up to 10 seconds) until the fees are displayed. In this case, the user can't submit the transaction, which is ok, but also, there is no notice of what's happening and why the fees are not displayed.
Check videos [1](https://drive.google.com/file/d/12RoYQAkyzSrcg6FVcbXumB7NwzGu-9tF/view?usp=sharing), [2](https://drive.google.com/file/d/1s1MOUbfhkV_eSH6-m8izYj6COpVHwM1x/view?usp=sharing) and [3](https://drive.google.com/file/d/1V1kOylIdiX9cWI-MBKTyzsbf0X44-jMq/view?usp=sharing) of this incident.

**Non issue**
1 - Overall, the previous fee didn't overwrite the latest one. However, my only concern is the fees appeared to be the same for 1,7,70 and 700 ADA. Does this have something to do with this PR? See [Test document](https://docs.google.com/spreadsheets/d/19BTZUQbL3ewoIYhqCttyB6d-19cv4b4LmpyjAyx0_Js/edit#gid=0)

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by mioriohk** [comment]
Hi @yakovkaravelov . This is the issue that I found on build 15336

**New issue**
1 -  I noticed an odd behaviour. I had a wallet with 21.8 ADA. Tried sending 7.8 ADA but the error message was saying I didn’t have enough ADA. The error on the console said I needed 123 ADA to proceed. Unable to replicate on production build 15062. [See recording here](https://drive.google.com/file/d/1aF9YQZWcsWNngsDFneFWVS6PuvFLHyYo/view?usp=sharing)

**Comment by yakovkaravelov** [comment]
@mioriohk Fixed that issue, please test again, thanks.

**Comment by mioriohk** [comment]
Hi @yakovkaravelov . Report on build 15378

**Fixed*
1 -  I noticed an odd behaviour. I had a wallet with 21.8 ADA. Tried sending 7.8 ADA but the error message was saying I didn’t have enough ADA. The error on the console said I needed 123 ADA to proceed. Unable to replicate on production build 15062. [See recording here](https://drive.google.com/file/d/1aF9YQZWcsWNngsDFneFWVS6PuvFLHyYo/view?usp=sharing)

**Review by mioriohk — APPROVED** [review]
LGTM 👍

**Comment by mioriohk** [comment]
Hi @yakovkaravelov . This is my findings for build 15398.

**New issues**
1 - It seems the fee is being overwritten by the previous amount's fee up to a point when it gets back to the correct one. I have prepared a document with screenshots. The entered amounts are executed consequentially. Please let me know if you have any question. Thank you. [Test document](https://docs.google.com/spreadsheets/d/19BTZUQbL3ewoIYhqCttyB6d-19cv4b4LmpyjAyx0_Js/edit#gid=0)

**Comment by mioriohk** [comment]
@yakovkaravelov As discussed, I have repeated the test on build 15398. Overall, the previous fee didn't overwrite the latest one. However, my only concern is the fees appeared to be the same for 1,7,70 and 700 ADA. Does this have something to do with this PR? See [Test document](https://docs.google.com/spreadsheets/d/19BTZUQbL3ewoIYhqCttyB6d-19cv4b4LmpyjAyx0_Js/edit#gid=0)

**Comment by yakovkaravelov** [comment]
> @yakovkaravelov As discussed, I have repeated the test on build 15398. Overall, the previous fee didn't overwrite the latest one. However, my only concern is the fees appeared to be the same for 1,7,70 and 700 ADA. Does this have something to do with this PR? See [Test document](https://docs.google.com/spreadsheets/d/19BTZUQbL3ewoIYhqCttyB6d-19cv4b4LmpyjAyx0_Js/edit#gid=0)

@mioriohk Cardano back-end api returns these same values, so this PR has nothing to do with these values.
CC @nikolaglumac

**Comment by mioriohk** [comment]
> @mioriohk Cardano back-end api returns these same values, so this PR has nothing to do with these values.
> CC @nikolaglumac

LGTM @yakovkaravelov , thank you 👍

**Comment by nikolaglumac** [comment]
> @yakovkaravelov As discussed, I have repeated the test on build 15398. Overall, the previous fee didn't overwrite the latest one. However, my only concern is the fees appeared to be the same for 1,7,70 and 700 ADA. Does this have something to do with this PR? See [Test document](https://docs.google.com/spreadsheets/d/19BTZUQbL3ewoIYhqCttyB6d-19cv4b4LmpyjAyx0_Js/edit#gid=0)

@mioriohk this depends on your wallet UTXO. The transaction amount plays no role in fee size. Fees will be the same if you have 1 Billion or 1 ADA on a single UTXO and that UTXO is being used to send 1 ADA.
Talk to @gnpf if this is unclear...

**Review by nikolaglumac — APPROVED** [review]
Works perfect now 🎉

**Comment by nikolaglumac** [comment]
@gnpf please test this one!

FYI @yakovkaravelov we won't be merging this until @gnpf approves 🙏

**Comment by gnpf** [comment]
@yakovkaravelov I find two issues for this PR:

1 - When switching the value, sometimes the "calculating fees" message disappears (for up to 10 seconds) until the fees are displayed. In this case, the user can't submit the transaction, which is ok, but also, there is no notice of what's happening and why the fees are not displayed.
Check videos [1](https://drive.google.com/file/d/12RoYQAkyzSrcg6FVcbXumB7NwzGu-9tF/view?usp=sharing), [2](https://drive.google.com/file/d/1s1MOUbfhkV_eSH6-m8izYj6COpVHwM1x/view?usp=sharing) and [3](https://drive.google.com/file/d/1V1kOylIdiX9cWI-MBKTyzsbf0X44-jMq/view?usp=sharing) of this incident.

2 - When the value changes, the "next" button is still enabled for a fraction of a second. If you quickly press enter after changing the value, you'll get to the confirmation dialog with no fees. Check[ this video.](https://drive.google.com/file/d/1nCl8MfzIgsnidmLab_gXS7m5Y1uJIWAo/view?usp=sharing)
Eventually, the correct fees are displayed in the dialog and the user can submit the transaction, but I'm wondering if this can be improved to avoid this awkward scenario.

**Comment by nikolaglumac** [comment]
@mioriohk @ManusMcCole can you please check if the [issues](https://github.com/input-output-hk/daedalus/pull/2239#issuecomment-738025154) @gnpf reported have been fixed in the latest version?

**Comment by ManusMcCole** [comment]
Hi @nikolaglumac. I can confirm that the issues reported by @gnpf have been resolved. Fee calculation is now super fast. See <a href="https://drive.google.com/file/d/17NXc_5NQg8HWpS6W1uaSY-18uqcNtK98/view?usp=sharing">video</a>. Goodjob @yakovkaravelov

**Review by ManusMcCole — APPROVED** [review]

**Comment by gnpf** [comment]
@yakovkaravelov Nice work, the fee calculation is very fast now so issue 1 is fixed. I can still reproduce the second issue in some scenarios:

> 2 - When the value changes, the "next" button is still enabled for a fraction of a second. If you quickly press enter after changing the value, you'll get to the confirmation dialog with no fees. Check this video.
Eventually, the correct fees are displayed in the dialog and the user can submit the transaction, but I'm wondering if this can be improved to avoid this awkward scenario.

If you enter a valid amount and then change it for an invalid amount (higher than the actual balance or below 1), there is a split second where you can press **Enter** and the "Send" button is still enabled. Check videos [1](https://drive.google.com/file/d/1lvFdh8r8wlh_ZecX1ug_OKCoHpXmOZfy/view?usp=sharing) and [2](https://drive.google.com/file/d/1nIbxQSCQLgrMyUolNtf8-sp3mEEmKaPE/view?usp=sharing). Can this be improved?

**Comment by yakovkaravelov** [comment]
> @yakovkaravelov Nice work, the fee calculation is very fast now so issue 1 is fixed. I can still reproduce the second issue in some scenarios:
> 
> > 2 - When the value changes, the "next" button is still enabled for a fraction of a second. If you quickly press enter after changing the value, you'll get to the confirmation dialog with no fees. Check this video.
> > Eventually, the correct fees are displayed in the dialog and the user can submit the transaction, but I'm wondering if this can be improved to avoid this awkward scenario.
> 
> If you enter a valid amount and then change it for an invalid amount (higher than the actual balance or below 1), there is a split second where you can press **Enter** and the "Send" button is still enabled. Check videos [1](https://drive.google.com/file/d/1lvFdh8r8wlh_ZecX1ug_OKCoHpXmOZfy/view?usp=sharing) and [2](https://drive.google.com/file/d/1nIbxQSCQLgrMyUolNtf8-sp3mEEmKaPE/view?usp=sharing). Can this be improved?

@gnpf This was fixed. Please test again. Thanks.

**Comment by gnpf** [comment]
@yakovkaravelov I can still reproduce the issue mentioned [in the previous comment](https://github.com/input-output-hk/daedalus/pull/2239#issuecomment-744415948). 
[Check this new video.](https://drive.google.com/file/d/1Q66eNBRQdqxvcPFJeniFN4xCJOWCglxM/view?usp=sharing)

Additionally, I got [this message](https://drive.google.com/file/d/1sARG5F7AHgazPPP2mbNeK9ODBdBvKBtf/view?usp=sharing) a few times. Not sure if it's an issue, but it was persistent the first time I installed this build. Then I cleaned the state and I couldn't reproduce it again.

**Comment by yakovkaravelov** [comment]
> @yakovkaravelov I can still reproduce the issue mentioned [in the previous comment](https://github.com/input-output-hk/daedalus/pull/2239#issuecomment-744415948).
> [Check this new video.](https://drive.google.com/file/d/1Q66eNBRQdqxvcPFJeniFN4xCJOWCglxM/view?usp=sharing)
> 
> Additionally, I got [this message](https://drive.google.com/file/d/1sARG5F7AHgazPPP2mbNeK9ODBdBvKBtf/view?usp=sharing) a few times. Not sure if it's an issue, but it was persistent the first time I installed this build. Then I cleaned the state and I couldn't reproduce it again.

@gnpf That is just short slot of time which react action handler processes to update state, and it can't be fixed.
And regarding new error message, that is not related to this ticket. It is about back-end api stuff.
Thanks.

**Comment by yakovkaravelov** [comment]
> @yakovkaravelov I can still reproduce the issue mentioned [in the previous comment](https://github.com/input-output-hk/daedalus/pull/2239#issuecomment-744415948).
> [Check this new video.](https://drive.google.com/file/d/1Q66eNBRQdqxvcPFJeniFN4xCJOWCglxM/view?usp=sharing)
> 
> Additionally, I got [this message](https://drive.google.com/file/d/1sARG5F7AHgazPPP2mbNeK9ODBdBvKBtf/view?usp=sharing) a few times. Not sure if it's an issue, but it was persistent the first time I installed this build. Then I cleaned the state and I couldn't reproduce it again.

Fixed. You can review again, @gnpf , thanks.

**Comment by nikolaglumac** [comment]
@darko-mijic please test this one. We won't be merging it without your approval!

**Comment by gnpf** [comment]
@yakovkaravelov The button is still enabled and the issue still reproduces [when the user clicks](https://drive.google.com/file/d/1vxiYgra7NSL4zEO69p-rjoehQJ3qTllp/view?usp=sharing). There is a delay when pressing "Enter" now, so it doesn't reproduce.
[This](https://drive.google.com/file/d/1So7Yemg4EhNanl1r8R5yNlKbTg56-YsX/view?usp=sharing) is how it looked on production, can't reproduce it there.

**Comment by nikolaglumac** [comment]
@gnpf @yakovkaravelov please have a call and finish this 🙏

**Review by gnpf — APPROVED** [review]
@yakovkaravelov Looks good now, great job

File: source/renderer/app/components/wallet/WalletSendForm.js
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:396]: @yakovkaravelov can we turn this line into a variable? That way it will be easier to understand what it means and we can reuse it on the 408 line...

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
Top work @yakovkaravelov!
I have left you one change request 🙏

**Comment by nikolaglumac** [comment]
@gnpf please review this one again 🙏

**Review by nikolaglumac — APPROVED** [review]
Excellent work @yakovkaravelov 💯

**Pull request merged**
